### PR TITLE
Chef bootstrap refactor and build JS packages on adhoc builds

### PR DESCRIPTION
**Pull request by wjordan**
This PR adds support for building javascript-package targets (apps, code-studio, blockly-core) from scratch on standalone-adhoc deployments (adding ~5min to the build time).

For now, we'll always build all javascript packages when running in local mode. In the future, perhaps we can optimize by conditionally building packages only when directory-specific changes (when compared against upstream `staging branch`) are detected.

I've also added a small bit of incremental cleanup to the Bootstrap Bash scripts.

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -83,6 +83,7 @@ namespace :build do
       RakeUtils.system './deploy.sh'
     end
   end
+  task :'blockly-core' => :blockly_core
```
Comment (Hamms) [inline Rakefile:86]: why is this necessary?

File: aws/chef-bootstrap.sh [hunk 2]
```diff
@@ -106,13 +107,13 @@ if [ -f /etc/chef/client.pem ] ; then
 fi
 
 if [ "${LOCAL_MODE}" = "1" ]; then
```
Comment (Hamms) [inline aws/chef-bootstrap.sh:110]: should this be updated to use the new `local_mode` variable?

File: aws/chef-bootstrap.sh [hunk 3]
```diff
@@ -106,13 +107,13 @@ if [ -f /etc/chef/client.pem ] ; then
...
-  mkdir -p /opt/chef-zero/{cookbooks,environments}
+  mkdir -p /var/chef/{cookbooks,environments}
```
Comment (Hamms) [inline aws/chef-bootstrap.sh]: likewise, should these be updated to use the new `chef_repo_path` variable?

File: cookbooks/cdo-apps/libraries/cdo_apps.rb [hunk 4]
```diff
@@ -104,4 +104,28 @@ def setup_app(app_name)
...
+      'LANG' => utf8,
+      'RAILS_ENV' => node.chef_environment
+    }
+    execute "build:#{package}" do
+      command "bundle exec rake build:#{package}"
+      cwd root
+      environment env.merge(node['cdo-apps']['bundle_env'])
+      user user
+      group user
+      not_if { File.directory? "#{root}/#{package}/node_modules"}
```
Comment (Hamms) [inline cookbooks/cdo-apps/libraries/cdo_apps.rb:127]: unfortunately, for code-studio we call the task `code_studio`, but the directory `code-studio`.

File: Rakefile (hunk 1 above)
Comment (Hamms) [inline Rakefile:86]: Oh wait, I see now.

We probably also want one for `"code-studio"` then, yeah?

File: lib/cdo/aws/s3_packaging.rb [hunk 5]
```diff
@@ -39,7 +39,7 @@ def update_from_s3
     begin
       ensure_updated_package
     rescue Aws::S3::Errors::NoSuchKey
-      @logger.info 'Package does not exist on S3. If you have made local changes to code-studio, you need to set build_code_studio and use_my_code_studio to true in locals.yml'
+      @logger.info "Package does not exist on S3. If you have made local changes to #{@package_name}, you need to set build_#{@package_name} and use_my_#{@package_name} to true in locals.yml"
```
Comment (Hamms) [inline lib/cdo/aws/s3_packaging.rb:42]: It's a bit unfortunate that `blockly-core` doesn't require a `use_my`, just a `build` ... it makes this line potentially misleading.

I'm not sure how worried I am about that, because setting a useless `use_my_blockly_core` isn't harmful.

**Comment by Hamms** [comment]
Couple of nits, and it looks like the Rakefile needs a `task :'code-studio' => :code_studio` to deal with the `-` vs `_` that exists in the chef recipes, but otherwise LGTM!

File: cookbooks/cdo-apps/libraries/cdo_apps.rb (hunk 4 above)
Comment (wjordan) [inline cookbooks/cdo-apps/libraries/cdo_apps.rb:127]: oh, good catch! Do you have any preference on adding the Rakefile aliases for the dash keys (as I've already done with blockly-core), versus just explicitly mapping the underscore folder to dashed versions of rake tasks within this function (e.g., via `package.tr('_', '-')`)?

Comment (Hamms) [inline cookbooks/cdo-apps/libraries/cdo_apps.rb:127]: I think I'd lean toward the Rakefile aliases; by mapping underscores to dashes elsewhere, we're assuming that the two are interchangable. By setting up aliases, we're declaring that they are.

File: aws/chef-bootstrap.sh (hunk 3 above)
Comment (wjordan) [inline aws/chef-bootstrap.sh]: there's a bit of admittedly-confusing [heredoc](http://tldp.org/LDP/abs/html/here-docs.html) language-switching going on above: this file is a Bash shell script that writes content (via inline heredoc syntax) to a `.rb` Ruby DSL (`chef_repo_path` and `local_mode` are Chef [`client.rb`](https://docs.chef.io/config_rb_client.html) parameters), that will be loaded/executed as Ruby during the initial `chef-client` run. So it's not possible to use the inner Ruby DSL parameters directly in the outer Bash script. What I can do at least is add a variable to clean up the `/var/chef` repetition, which should help a bit.

Comment (Hamms) [inline aws/chef-bootstrap.sh]: hah, that pattern is familiar from my time spent in chef land. Makes sense to me!
